### PR TITLE
Make the crafting UI look like the vanilla crafting table

### PR DESCRIPTION
**Pull request by viciscat**
Self explanatory ain't it?

<details>
<summary>High Res Images</summary>

image

image

</details>

### CODE REVIEW

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableHandler.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+import net.minecraft.entity.player.PlayerInventory;
+import net.minecraft.inventory.Inventory;
+import net.minecraft.screen.GenericContainerScreenHandler;
+import net.minecraft.screen.ScreenHandlerType;
+import net.minecraft.screen.slot.Slot;
+
+import java.util.Arrays;
+
+public class SkyblockCraftingTableHandler extends GenericContainerScreenHandler {
```
Comment (kevinthegreat1) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableHandler.java:11]:
```suggestion
public class SkyblockCraftingTableScreenHandler extends GenericContainerScreenHandler {
```

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreen.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    @Override
+    protected void onMouseClick(Slot slot, int slotId, int button, SlotActionType actionType) {
+        super.onMouseClick(slot, slotId, button, actionType);
+        this.recipeBook.slotClicked(slot);
+        // Don't make the more item head skull thing show up
+        if (slot != null && slot.id == 26) handler.setCursorStack(ItemStack.EMPTY);
+    }
+
+
+    static class Dummy extends AbstractRecipeScreenHandler<SimpleInventory> {
```
Comment (kevinthegreat1) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreen.java:147]:
```suggestion
    static class DummyRecipeScreenHandler extends AbstractRecipeScreenHandler<SimpleInventory> {
```

**Review by kevinthegreat1 — CHANGES_REQUESTED** [review]
Code looks good, some naming convensions, and needs testing.

**Comment by x-yingcan-x** [comment]
pls remove the `quick craft` in texture, we all know what it is and this will only make it worse in I18N support (hard to translate)

**Comment by viciscat** [comment]
> pls remove the `quick craft` in texture, we all know what it is and this will only make it worse in I18N support (hard to translate)
Not in the texture, however the string is hard coded. Imma change that

**Comment by Julienraptor01** [comment]
> pls remove the `quick craft` in texture, we all know what it is and this will only make it worse in I18N support (hard to translate)

I kinda agree with the fact the quick craft line is kinda bad ?
How does Hypixel communicate that these slots are quick-craft ?
If they don't put any mention we probably should neither

**Comment by viciscat** [comment]
tooltip is already there cuz it literally just displays the items in that slot. So I'm gonna yeet the string

**Comment by Julienraptor01** [comment]
> tooltip is already there cuz it literally just displays the items in that slot. So I'm gonna yeet the string

it displays the glass panes ?
and also display the barrier blocks ?
would there be a way to show these as empty, but still show the tooltips
and also obv make sure that putting items in those slots behave as vanilla / is impossible (which is probably easier to do)
did you prevent the players from even switching items from the cursor slot to the output slot ?

**Comment by viciscat** [comment]
I made barrier not render cuz it looked weird, but glass panes still render. They don't look too out of place

**Comment by Julienraptor01** [comment]
> I made barrier not render cuz it looked weird, but glass panes still render. They don't look too out of place
> 
> EDIT: did not prevent clicking on output slot aka barrier, and base skyblocker prevents clicking on black stained glass panes

actually preventing to click could go wrong as if you click with the same item, it does things, so not preventing the output one is fine
can you show an example with quickcraft slots with panes ?

**Comment by viciscat** [comment]
~~test it yourself~~

Here's when ya got no crafts
<details>
<summary>In spoiler cuz these images real big</summary>

image

Simple hover for More Crafts
image


</details>

**Comment by Julienraptor01** [comment]
> ~~test it yourself~~
> 
> Here's when ya got no crafts
> <details>
> <summary>In spoiler cuz these images real big</summary>
> 
> image
> 
> Simple hover for More Crafts
> image
> 
> 
> </details>

Is the more craft tooltip hardcoded ?
Or do you extract info from the tooltips of the + head item ?

**Comment by viciscat** [comment]
hard coded, I can try and get the original tooltip

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreen.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    protected void init() {
+        super.init();
+        this.narrow = this.width < 379;
+        this.recipeBook.initialize(this.width, this.height, this.client, this.narrow, new DummyRecipeScreenHandler());
+        this.x = this.recipeBook.findLeftEdge(this.width, this.backgroundWidth) + 11;
+        this.addDrawableChild(new TexturedButtonWidget(this.x + 5, this.height / 2 - 49, 20, 18, RecipeBookWidget.BUTTON_TEXTURES, button -> {
+            this.recipeBook.toggleOpen();
+            this.x = this.recipeBook.findLeftEdge(this.width, this.backgroundWidth) + 11;
+            button.setPosition(this.x + 5, this.height / 2 - 49);
+            if(moreCraftsButton != null) moreCraftsButton.setPosition(this.x + 174, this.y+62);
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreen.java:54]: format

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreen.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        if (slot.id == 23 && slot.getStack().isOf(Items.BARRIER)) return;
+        super.drawSlot(context, slot);
+    }
+
+    @Override
+    protected void drawBackground(DrawContext context, float delta, int mouseX, int mouseY) {
+        int i = this.x;
+        int j = (this.height - this.backgroundHeight) / 2;
+        context.drawTexture(TEXTURE, i, j, 0, 0, this.backgroundWidth, this.backgroundHeight);
+        context.drawGuiTexture(QUICK_CRAFT, i+173, j, 0, 25, 84);
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreen.java:103]: format

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreen.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        }
+        if (this.narrow && this.recipeBook.isOpen()) {
+            return true;
+        }
+        return super.mouseClicked(mouseX, mouseY, button);
+    }
+
+    @Override
+    protected boolean isClickOutsideBounds(double mouseX, double mouseY, int left, int top, int button) {
+        boolean bl = mouseX < (double)left || mouseY < (double)top || mouseX >= (double)(left + this.backgroundWidth) || mouseY >= (double)(top + this.backgroundHeight);
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreen.java:125]: add spaces between casts

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreenHandler.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+public class SkyblockCraftingTableScreenHandler extends GenericContainerScreenHandler {
+
+    private static final int[] normalSlots = new int[]{
+            10, 11, 12,     16,
+            19, 20, 21, 23, 25,
+            28, 29, 30,     34
+    };
+    public SkyblockCraftingTableScreenHandler(ScreenHandlerType<?> type, int syncId, PlayerInventory playerInventory, Inventory inventory, int rows) {
+        super(type, syncId, playerInventory, inventory, rows);
+        for (int i = 0; i < rows*9; i++) {
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreenHandler.java:20]: format

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreenHandler.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                slot.id = i;
+                slots.set(i, slot);
+            } else {
+                DisabledSlot slot = new DisabledSlot(originalSlot.inventory, originalSlot.getIndex(), originalSlot.x, originalSlot.y);
+                slot.id = i;
+                slots.set(i, slot);
+            }
+        }
+        int yOffset = (rows - 4) * 18 + 19;
+        for (int i = rows*9; i < slots.size(); i++) {
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreenHandler.java:34]: format

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreenHandler.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    public SkyblockCraftingTableScreenHandler(GenericContainerScreenHandler handler, PlayerInventory playerInventory) {
+        this(handler.getType(), handler.syncId, playerInventory, handler.getInventory(), handler.getRows());
+    }
+
+    private int[] getCoords(int slot) {
+        if (slot == 23) return new int[]{124, 35};
+        if (slot == 16 || slot == 25 || slot == 34) {
+            int y = (slot/9 -  1) * 18 + 8;
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreenHandler.java:49]: format

File: src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreenHandler.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            return new int[]{174, y};
+        }
+        int gridX = slot%9 - 1;
+        int gridY = slot/9 - 1;
+        return new int[]{30 + gridX*18, 17+gridY*18};
```
Comment (AzureAaron) [inline src/main/java/de/hysky/skyblocker/skyblock/item/SkyblockCraftingTableScreenHandler.java:54]: format 3x

**Review by AzureAaron — CHANGES_REQUESTED** [review]

**Review by kevinthegreat1 — CHANGES_REQUESTED** [review]
Forgot to add config. Also this doesn't work if `betterPartyFinder` is turned off. See `HandledScreenProviderMixin`.

**Review by kevinthegreat1 — APPROVED** [review]

**Pull request merged**